File: python/course-repo/solution/02-idempotency-and-crash-safety/src/webhooks/activities.py
import httpx
from temporalio import activity
from temporalio.exceptions import ApplicationError

from .shared import WebhookDelivery


@activity.defn
def deliver_webhook(req: WebhookDelivery) -> int:
    info = activity.info()
    activity.logger.info(
        "Delivering webhook for event %s (attempt %d)", req.event_id, info.attempt
    )

    # The webhook event id is stable across retries, so every retry POSTs the
    # same logical delivery key and the receiver can dedupe the side effect.
    headers = {"Idempotency-Key": f"webhook:{req.event_id}"}

    response = httpx.post(req.url, json=req.payload, headers=headers, timeout=10.0)
    response.raise_for_status()

    # Simulate transient failure on attempts 1-2. ApplicationError defaults to
    # retryable; set non_retryable=True for permanent failures.
    if info.attempt < 3:
        raise ApplicationError(
            f"Simulated transient failure on attempt {info.attempt}",
        )

    return response.status_code

File: typescript/course-repo/solution/02-idempotency-and-crash-safety/src/activities.ts
import { ApplicationFailure, activityInfo, log } from '@temporalio/activity';
import type { WebhookDelivery } from './shared';

export async function deliverWebhook(req: WebhookDelivery): Promise<number> {
  const { attempt } = activityInfo();
  log.info('Delivering webhook for event', { eventId: req.eventId, attempt });

  // The webhook event id is stable across retries, so every retry POSTs the
  // same logical delivery key and the receiver can dedupe the side effect.
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    'Idempotency-Key': `webhook:${req.eventId}`,
  };

  const response = await fetch(req.url, {
    method: 'POST',
    headers,
    body: JSON.stringify(req.payload),
  });
  if (!response.ok) throw new Error(`HTTP ${response.status}`);

  // Simulate transient failure on attempts 1-2.
  if (attempt < 3) {
    throw ApplicationFailure.create({
      message: `Simulated transient failure on attempt ${attempt}`,
    });
  }

  return response.status;
}
